base/expr.jl

```diff
@@ -371,6 +371,187 @@ macro constprop(setting, ex)
     throw(ArgumentError("@constprop $setting not supported"))
 end
 
+"""
+    @assume_effects setting... ex
+    @assume_effects(setting..., ex)
+
+`@assume_effects` overrides the compiler's effect modeling for the given method.
+`ex` must be a method definition or `@ccall` expression.
+
+WARNING: Improper use of this macro causes undefined behavior (including crashes,
+incorrect answers, or other hard to track bugs). Use with care and only if absolutely
+required.
+
+In general, each `setting` value makes an assertion about the behavior of the
```

Comment:
I think the rest of this docstring might be more appropriate to include directly in the devdocs, with a link from here, rather than directly here?

Reply:
It seems appropriate to have it here, since users may want to refer to the list of conditions when they're thinking about applying the macro. Devdocs feel more like internal details that users would never have to worry about.